### ISSUE TITLE
Using internal include files across packages (UMFPACK uses amd_internal.h, etc)

### ISSUE THREAD

**Issue by DrTimothyAldenDavis** [DrTimothyAldenDavis/SuiteSparse#466]
Some packages in SuiteSparse use internal #include files from other packages in SuiteSparse.  UMFPACK uses amd_internal.h, for example.  This makes building the libraries awkward.

The short-term workaround was to copy amd_internal.h into UMFPACK/Source/, but this is not a good long-term solution.

A better mechanism would be to fold the contents of amd_internal.h into the public header amd.h, but to do so in a way that ensures the definitions are not exposed to a normal external package.  UMFPACK would use

```
#define AMD__USE_INTERNAL_DEFINITIONS
#include "amd.h"
```

and then the contents of amd_internal.h would appended to amd.h with

```
// amd.h existing contents here
#ifdef AMD__USE_INTERNAL_DEFINITIONS
// contents of amd_internal.h goes here
#endif
#undef AMD__USE_INTERNAL_DEFINITIONS
```

Then the amd_internal.h file would simply be deleted.  The same issue occurs with Paru, which #includes umf_internal.h.

**Comment by mmuetzel** [comment]
That sounds like a reasonable solution to me.
In case it is only a subset of functions currently declared in the `*_internal.h` headers that are needed by other libraries, it might also be reasonable to only move that subset to the installed header. Probably still guarded by a new preprocessor variable.

**Comment by DrTimothyAldenDavis** [comment]
Reviewing the whole code base of SuiteSparse, this issue occurs in two places, when compiling the primary libraries.  These 2 cases must be fixed:

(1) UMFPACK requires amd_internal.h.  This will be easy to fix.
(2) ParU requires UMFPACK/Source/umf_internal.h and thus also requiring amd_internal.h, umf_version.h, umf_config.h, umf_dump.h.  All the umf_*.h files are in the UMFPACK/Source folder, and are not public headers.  This will be a little harder to fix, but it won't be unreasonable.

There are other places where internal include files, and even source files, are used, but this is safe.  CHOLMOD/Tcov is the full statement coverage test for CHOLMOD, AMD, CAMD, COLAMD, and CCOLAMD, and this it requires the entire source code for all those packages.  So I can't revise that.  But the Tcov tests are used by the cmake build system and don't appear in the final compiled libraries.  They require the whole SuiteSparse source code base.  So there's nothing to fix.

**Comment by DrTimothyAldenDavis** [comment]
This is now fixed in UMFPACK for SuiteSparse 7.4.0, in the default dev branch of the SuiteSparse repo.  I'll release a 7.4.0.beta1 soon.

It's not yet fixed in ParU, however, so I'll leave this issue open.  I've tagged ParU at version 0.1.0, with this and a few other minor issues, and will release ParU 0.1.0 in SuiteSparse 7.4.0.

**Comment by DrTimothyAldenDavis** [comment]
I'll close this, and leave ParU with a "FIXME" in the code.
